### PR TITLE
feature/datatrans security update

### PR DESCRIPTION
**Pull request by ludmilablum**
BasicAuth and hmac to comply with Datatrans minimal security requirements

### CODE REVIEW

File: Client/Authentication/AuthenticationStrategy.php [hunk 1]
```diff
@@ -1,98 +1,137 @@
...
-     * @var string
-     */
+    /** @var string */
     protected $sign;
 
+    /** @var string */
+    protected $httpAuthUserPwd;
+
+    /** @var string */
+    protected $hmacKey;
```
Comment (starkj) [inline Client/Authentication/AuthenticationStrategy.php]: `$hmacKey` seems not to be used anywhere expect in `setHmacKey` where it is set, no further use there either

File: Client/Client.php [hunk 2]
```diff
@@ -140,77 +119,52 @@ protected function fillDataFromXML(SettlementResponse $settlementResponse, $xml)
...
-            } else {
-                $postFields = $request->request->all();
-            }
+        $this->logger->info($parameter->getRequestUrl());
+        $this->logger->info("Request:\n" . var_export($data, true));
+
+        $curlOpts = [
+            CURLOPT_URL => $parameter->getRequestUrl(),
+            CURLOPT_POST => true,
+            CURLOPT_SSL_VERIFYHOST => false,
```
Comment (starkj) [inline Client/Client.php]: should be set to `2` according to https://www.php.net/manual/en/function.curl-setopt.php

File: Client/Client.php [hunk 3]
```diff
@@ -140,77 +119,52 @@ protected function fillDataFromXML(SettlementResponse $settlementResponse, $xml)
...
+            CURLOPT_SSL_VERIFYPEER => false,
```
Comment (starkj) [inline Client/Client.php]: should be set to `true`

File: Plugin/DatatransPlugin.php [hunk 4]
```diff
@@ -87,12 +88,22 @@ public function approve(FinancialTransactionInterface $transaction, $retry)
...
 
         $request = $this->requestStack->getMasterRequest();
-        if ($request !== null && $request->request->has('responseCode') ) {
+        if ($request !== null
+            && $request->request->has(PayConfirmParameter::PAY_PARAM_RESPONSECODE)) {
+
+            if ($request->request->get(PayConfirmParameter::PAY_PARAM_RESPONSECODE)
+                != PayConfirmParameter::PAY_PARAM_RESPONSECODE_SUCCESS) {
+
+                $payConfirmParameter = $this->client->getConfirmParameter($request->request);
```
Comment (starkj) [inline Plugin/DatatransPlugin.php]: Why call `getConfirmParameter()` and assign the result to `$payConfirmParameter` if en exception gets thrown by the next line?

File: Plugin/PayConfirmParameter.php [hunk 5]
```diff
@@ -1,163 +1,189 @@
...
+    public const PAY_INIT_PARAM_SUCCESS_URL = 'successUrl';
+    public const PAY_PARAM_AMOUNT = 'amount';
+    public const PAY_PARAM_CURRENCY = 'currency';
+    public const PAY_PARAM_ERRORCODE = 'errorCode';
+    public const PAY_PARAM_ERRORDETAIL = 'errorDetail';
+    public const PAY_PARAM_ERRORMESSAGE = 'errorMessage';
+    public const PAY_PARAM_REFNO = 'refno';
+    public const PAY_PARAM_REQTYPE = 'reqtype';
+    public const PAY_PARAM_RESPONSECODE = 'responseCode';
+    public const PAY_PARAM_RESPONSECODE_SUCCESS = 01;
```
Comment (starkj) [inline Plugin/PayConfirmParameter.php]:
```suggestion
    public const PAY_PARAM_RESPONSECODE_SUCCESS = '01';
```

File: Plugin/DatatransPlugin.php [hunk 6]
```diff
@@ -213,7 +227,12 @@ protected function throwUnlessSuccessPayComplete(SettlementResponse $settlementR
      */
     protected function createSettlementRequest(FinancialTransactionInterface $transaction)
     {
-        $referenceNumber = $transaction->getPayment()->getApproveTransaction()->getTrackingId();
+        $approveTransaction = $transaction->getPayment()->getApproveTransaction();
+        if ($approveTransaction == null) {
```
Comment (starkj) [inline Plugin/DatatransPlugin.php]:
```suggestion
        if ($approveTransaction === null) {
```

File: Client/Authentication/AuthenticationStrategy.php [hunk 7]
```diff
@@ -1,98 +1,137 @@
...
-     * @var string
-     */
+    /** @var string */
     protected $merchantId;
 
-    /**
-     * @var string
-     */
+    /** @var string */
     protected $sign;
```
Comment (starkj) [inline Client/Authentication/AuthenticationStrategy.php]: What about calling this attribute `merchantPwd`?

File: Plugin/DatatransPlugin.php [hunk 8]
```diff
@@ -127,19 +135,15 @@ public function deposit(FinancialTransactionInterface $transaction, $retry)
...
             $settlementResponse = $this->client->payComplete($settlementRequest);
             $this->throwUnlessSuccessPayComplete($settlementResponse);
-
-        } catch(\Exception $e) {
-            $this->throwFinancialTransaction($transaction, $e->getMessage());
+        } catch (Exception $e) {
+            $this->throwFinancialTransaction($transaction, $e->getMessage(), $settlementResponse->getResponseCode());
         }
 
-        $transaction->setReferenceNumber($settlementResponse->getRefno());
```
Comment (starkj) [inline Plugin/DatatransPlugin.php:138]: Was this line deleted on purpose? No `Refno` for settlements?

Comment (ludmilablum) [reply Plugin/DatatransPlugin.php:138]: Yes! It has already been set in the previous transaction and will not be provided in the settlementResponse again. Would be nulled here.

**Review by starkj — APPROVED** [review]

**Pull request merged**